contracts/multisig-prover/src/encoding/mvx/mod.rs

```diff
@@ -0,0 +1,438 @@
+        }
+    }
+}
+
+impl From<&VerifierSet> for WeightedSigners {
+    fn from(verifier_set: &VerifierSet) -> Self {
+        let mut signers = verifier_set
+            .signers
+            .values()
+            .map(WeightedSigner::from)
+            .collect::<Vec<_>>();
+
+        signers.sort_by_key(|weighted_signer| weighted_signer.signer);
+
+        WeightedSigners {
+            signers,
+            threshold: uint256_to_compact_vec(verifier_set.threshold.into()),
+            nonce: Keccak256::digest(Uint256::from(verifier_set.created_at).to_be_bytes()).into(),
+        }
+    }
+}
+
+impl TryFrom<&RouterMessage> for Message {
+    type Error = ContractError;
+
+    fn try_from(msg: &RouterMessage) -> Result<Self, Self::Error> {
+        let map_addr_err = |_| ContractError::InvalidMessage {
+            reason: format!(
+                "destination_address is not a valid MVX address: {}",
+                msg.destination_address.as_str()
+            ),
+        };
+
+        let (_, data, _) =
+            bech32::decode(&msg.destination_address.as_str()).map_err(map_addr_err)?;
+        let addr_vec = Vec::<u8>::from_base32(&data).map_err(map_addr_err)?;
+        let contract_address =
+            <[u8; 32]>::try_from(addr_vec).map_err(|_| ContractError::InvalidMessage {
+                reason: format!(
+                    "destination_address is not a valid MVX address: {}",
+                    msg.destination_address.as_str()
+                ),
+            })?;
+
+        Ok(Message {
+            source_chain: msg.cc_id.chain.to_string(),
+            message_id: msg.cc_id.id.to_string(),
+            source_address: msg.source_address.to_string(),
+            contract_address,
+            payload_hash: msg.payload_hash,
+        })
+    }
+}
+
+fn uint256_to_compact_vec(value: Uint256) -> Vec<u8> {
+    if value.is_zero() {
+        return Vec::new();
+    }
+
+    let bytes = value.to_be_bytes();
+    let mut slice_from = 0;
+    for (i, byte) in bytes.iter().enumerate() {
+        if *byte != 0 {
+            slice_from = i;
+            break;
+        }
+    }
+
+    bytes[slice_from..].to_vec()
+}
+
+pub fn ed25519_key(pub_key: &PublicKey) -> Result<[u8; 32], ContractError> {
+    match pub_key {
+        PublicKey::Ed25519(ed25519_key) => {
+            return Ok(<[u8; 32]>::try_from(ed25519_key.as_ref())
+                .expect("couldn't convert pubkey to ed25519 public key"));
+        }
+        _ => {
+            return Err(ContractError::InvalidPublicKey {
+                reason: "Public key is not ed25519".into(),
+            })
+        }
+    }
+}
+
+pub fn payload_hash_to_sign(
+    domain_separator: &Hash,
+    signer: &VerifierSet,
+    payload: &Payload,
+) -> Result<Hash, ContractError> {
+    let signer_hash = WeightedSigners::from(signer).hash();
+    let data_hash = Keccak256::digest(encode(payload)?);
+
+    let unsigned = [
+        "\x19MultiversX Signed Message:\n".as_bytes(),

```

Comment:
On MultiversX this is not required at all, but I just added it to be similar to EVM.

In our Gateway contract we have this as bytes constant also. I wouldn't do any changes here since I would have to also modify the contracts which are in audit atm.